Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+        }
+      case .clientOpenServerOpen(var state):
+        try state.deframer.process(buffer: message) { deframedMessage in
+          state.inboundMessageBuffer.append(deframedMessage)
+        }
+        
+        if endStream {
+          self.state = .clientClosedServerOpen(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerOpen(state)
+        }
+      case .clientOpenServerClosed:
+        // Client is not done sending request, but server has already closed.
+        // Ignore the rest of the request: do nothing.
+        ()
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client can't send a message if closed.")
+      }
+    }
+    
+    mutating func nextOutboundMessage() throws -> ByteBuffer? {
+      switch self.state {
+      case .clientIdleServerIdle, .clientOpenServerIdle, .clientClosedServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Server is not open yet.")
+      case .clientOpenServerOpen(var state):
+        let response = try state.framer.next(compressor: state.compressor)
+        self.state = .clientOpenServerOpen(state)
+        return response
+      case .clientClosedServerOpen:
+        // No point in sending response if client is closed: do nothing.
+        return nil
+      case .clientOpenServerClosed, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't send response if server is closed.")
+      }
+    }
+    
+    mutating func nextInboundMessage() -> [UInt8]? {
+      switch self.state {
+      case .clientOpenServerIdle(var state):
+        let request = state.inboundMessageBuffer.pop()
+        self.state = .clientOpenServerIdle(state)
+        return request
+      case .clientOpenServerOpen(var state):
+        let request = state.inboundMessageBuffer.pop()
+        self.state = .clientOpenServerOpen(state)
+        return request
+      case .clientClosedServerOpen(var state):
+        let request = state.inboundMessageBuffer.pop()
+        self.state = .clientClosedServerOpen(state)
+        return request
+      case .clientClosedServerIdle,
+          .clientIdleServerIdle,
+          .clientOpenServerClosed,
+          .clientClosedServerClosed:
+        return nil
+      }
+    }
+    
+    private func assertionFailureAndCreateRPCErrorOnFailedPrecondition(_ message: String) -> RPCError {
+      if !self.skipAssertions {
+        assertionFailure(message)
+      }
+      return RPCError(code: .failedPrecondition, message: message)
+    }
+  }
+}
+
+extension MethodDescriptor {
+  init?(fullyQualifiedMethod: String) {
+    let split = fullyQualifiedMethod.split(separator: "/")
+    guard split.count == 2 else {
+      return nil
+    }
+    self.init(service: String(split[0]), method: String(split[1]))
+  }
+}
+
+extension Metadata {
```

Comment:
These will actually need to be on `HPACKHeaders` and we should keep them internal

Reply:
We should remove this extension now